### PR TITLE
[AUDIO]

### PR DESCRIPTION
**Pull request by janderwald**
- don't load sysaudio and wdmaud from audio service. They are started automatically by SWBUS driver (swenum.sys)
- don't create device object with sysaudio with predefined name
- enumerate sysaudio devices dynamically
- this should fix several audio issues and audio now works again after first reboot

### CODE REVIEW

File: drivers/wdm/audio/legacy/wdmaud/deviface.c [hunk 1]
```diff
@@ -49,10 +73,58 @@ DeviceInterfaceChangeCallback(
...
+        Entry = (SYSAUDIO_ENTRY*)AllocateItem(NonPagedPool, sizeof(SYSAUDIO_ENTRY) + (Length + 1)* sizeof(WCHAR));
+        if (!Entry)
+        {
+            return STATUS_INSUFFICIENT_RESOURCES;
+        }
+
+        Entry->SymbolicLink.Length = Length * sizeof(WCHAR);
+        Entry->SymbolicLink.MaximumLength = Length * sizeof(WCHAR);
+        Entry->SymbolicLink.Buffer = (LPWSTR) (Entry + 1);
+        RtlCopyMemory(Entry->SymbolicLink.Buffer, Event->SymbolicLinkName->Buffer, Length);
```
Comment (ThFabba) [inline drivers/wdm/audio/legacy/wdmaud/deviface.c]: RtlInitEmptyUnicodeString + RtlCopyUnicodeString seems nicer, and will add a null terminator for you.

File: drivers/wdm/audio/legacy/wdmaud/deviface.c [hunk 2]
```diff
@@ -49,10 +73,58 @@ DeviceInterfaceChangeCallback(
...
+
+        if (!IsEntryAlreadyInList(DeviceExtension, Entry)) {
```
Comment (ThFabba) [inline drivers/wdm/audio/legacy/wdmaud/deviface.c]: Could make this function just take the PCUNICODE_STRING as a parameter instead of the whole entry, then you could do this check before the allocation and just use `&Event->SymbolicLinkName`

File: drivers/wdm/audio/legacy/wdmaud/mmixer.c [hunk 3]
```diff
@@ -793,8 +794,11 @@ CreatePinCallback(
     InstanceInfo.Flags = 0;
     InstanceInfo.DeviceNumber = VirtualDeviceId;
 
+    // FIXME currently sysaudio contains all ks devices
+    DeviceEntry = (PSYSAUDIO_ENTRY)CONTAINING_RECORD(Context->DeviceExtension->SysAudioDeviceList.Flink, SYSAUDIO_ENTRY, Entry);
```
Comment (ThFabba) [inline drivers/wdm/audio/legacy/wdmaud/mmixer.c]: Don't need the cast.

File: drivers/wdm/audio/legacy/wdmaud/sup.c [hunk 4]
```diff
@@ -54,11 +56,21 @@ GetSysAudioDeviceCount(
...
 
-    /* query sysaudio for the device count */
-    Status = KsSynchronousIoControlDevice(DeviceExtension->FileObject, KernelMode, IOCTL_KS_PROPERTY, (PVOID)&Pin, sizeof(KSPROPERTY), (PVOID)&Count, sizeof(ULONG), &BytesReturned);
-    if (!NT_SUCCESS(Status))
-        return 0;
+    Count = 0;
+    Entry = DeviceExtension->SysAudioDeviceList.Flink;
+    while (Entry != &DeviceExtension->SysAudioDeviceList)
+    {
+        DeviceEntry = (PSYSAUDIO_ENTRY)CONTAINING_RECORD(Entry, SYSAUDIO_ENTRY, Entry);
```
Comment (ThFabba) [inline drivers/wdm/audio/legacy/wdmaud/sup.c]: Don't need the cast.

File: drivers/wdm/audio/legacy/wdmaud/sup.c [hunk 5]
```diff
@@ -54,11 +56,21 @@ GetSysAudioDeviceCount(
...
 
+        /* query sysaudio for the device count */
+        Status = KsSynchronousIoControlDevice(DeviceEntry->FileObject, KernelMode, IOCTL_KS_PROPERTY, (PVOID)&Pin, sizeof(KSPROPERTY), (PVOID)&DeviceCount, sizeof(ULONG), &BytesReturned);
+        if (!NT_SUCCESS(Status)) {
+            DPRINT1("KSPROPERTY_SYSAUDIO_DEVICE_COUNT failed with %x\n", Status);
+            return 0;
+        }
+        Count += DeviceCount;
+		Entry = Entry->Flink;
```
Comment (ThFabba) [inline drivers/wdm/audio/legacy/wdmaud/sup.c]: Broken indentation.

File: drivers/wdm/audio/legacy/wdmaud/deviface.c [hunk 6]
```diff
@@ -12,6 +12,30 @@
...
+
+    PLIST_ENTRY Entry;
+    PSYSAUDIO_ENTRY DeviceEntry;
+
+    Entry = DeviceExtension->SysAudioDeviceList.Flink;
+    while (Entry != &DeviceExtension->SysAudioDeviceList)
+    {
+        DeviceEntry = (PSYSAUDIO_ENTRY)CONTAINING_RECORD(Entry, SYSAUDIO_ENTRY, Entry);
+
+        if (RtlEqualUnicodeString(&DeviceEntry->SymbolicLink, &NewDeviceEntry->SymbolicLink, TRUE)) {
```
Comment (HBelusca) [inline drivers/wdm/audio/legacy/wdmaud/deviface.c]: Braces not needed.

File: drivers/wdm/audio/legacy/wdmaud/deviface.c [hunk 7]
```diff
@@ -12,6 +12,30 @@
...
+            return TRUE;
+        }
+        Entry = Entry->Flink;
+    }
+	return FALSE;
```
Comment (HBelusca) [inline drivers/wdm/audio/legacy/wdmaud/deviface.c]: Indentation broken. Uses 4-spaces, not tabs!

File: drivers/wdm/audio/legacy/wdmaud/deviface.c [hunk 8]
```diff
@@ -12,6 +12,30 @@
...
+
+BOOLEAN
+IsEntryAlreadyInList(
+    IN PWDMAUD_DEVICE_EXTENSION DeviceExtension,
+    IN PSYSAUDIO_ENTRY NewDeviceEntry) {
+
+    PLIST_ENTRY Entry;
+    PSYSAUDIO_ENTRY DeviceEntry;
+
+    Entry = DeviceExtension->SysAudioDeviceList.Flink;
```
Comment (HBelusca) [inline drivers/wdm/audio/legacy/wdmaud/deviface.c]: Optional: You can use a for-loop.

File: drivers/wdm/audio/legacy/wdmaud/deviface.c (hunk 2 above)
Comment (HBelusca) [inline drivers/wdm/audio/legacy/wdmaud/deviface.c]: Curly brackets on their own lines please.

File: drivers/wdm/audio/legacy/wdmaud/deviface.c [hunk 9]
```diff
@@ -80,9 +151,34 @@ WdmAudOpenSysAudioDeviceInterfaces(
         Entry->SymbolicLink.Buffer = (LPWSTR) (Entry + 1);
         wcscpy(Entry->SymbolicLink.Buffer, SymbolicLinkList);
 
-        InsertTailList(&DeviceExtension->SysAudioDeviceList, &Entry->Entry);
+        if (!IsEntryAlreadyInList(DeviceExtension, Entry)) {
```
Comment (HBelusca) [inline drivers/wdm/audio/legacy/wdmaud/deviface.c]: Curly brackets on their own lines.

File: drivers/wdm/audio/legacy/wdmaud/deviface.c [hunk 10]
```diff
@@ -80,9 +151,34 @@ WdmAudOpenSysAudioDeviceInterfaces(
...
+                FreeItem(Entry);
+                return Status;
+            }
+
+            DPRINT1("SysaudioDevice %S\n", SymbolicLinkList);
+
+            InsertTailList(&DeviceExtension->SysAudioDeviceList, &Entry->Entry);
+            DeviceExtension->NumSysAudioDevices++;
+
+        } else {
```
Comment (HBelusca) [inline drivers/wdm/audio/legacy/wdmaud/deviface.c]: Ditto.

File: drivers/wdm/audio/legacy/wdmaud/sup.c [hunk 11]
```diff
@@ -54,11 +56,21 @@ GetSysAudioDeviceCount(
...
-        return 0;
+    Count = 0;
+    Entry = DeviceExtension->SysAudioDeviceList.Flink;
+    while (Entry != &DeviceExtension->SysAudioDeviceList)
+    {
+        DeviceEntry = (PSYSAUDIO_ENTRY)CONTAINING_RECORD(Entry, SYSAUDIO_ENTRY, Entry);
 
+        /* query sysaudio for the device count */
+        Status = KsSynchronousIoControlDevice(DeviceEntry->FileObject, KernelMode, IOCTL_KS_PROPERTY, (PVOID)&Pin, sizeof(KSPROPERTY), (PVOID)&DeviceCount, sizeof(ULONG), &BytesReturned);
+        if (!NT_SUCCESS(Status)) {
```
Comment (HBelusca) [inline drivers/wdm/audio/legacy/wdmaud/sup.c]: Curly bracket on their own lines.

**Comment by binarymaster** [comment]
> What did you change in the ReactOS_LogOn.wav file ??

It got resampled from 44.1 kHz to 48 kHz.

**Comment by HBelusca** [comment]
> 
> 
> > What did you change in the ReactOS_LogOn.wav file ??
> 
> It got resampled from 44.1 kHz to 48 kHz.

Question: why?

**Comment by janderwald** [comment]
> 
> 
> What did you change in the ReactOS_LogOn.wav file ??
I changed the sample rate from 44.1khz to 48khz because VirtualBox AC97 driver wont play with 44.1khz

**Comment by BenNottelling** [comment]
> > I changed the sample rate from 44.1khz to 48khz because VirtualBox AC97 driver wont play with 44.1khz

Really? because it will occasionally play it

**Comment by Extravert-ir** [comment]
GCC build error:
```
../drivers/wdm/audio/legacy/wdmaud/entry.c: In function 'WdmAudCreate':
../drivers/wdm/audio/legacy/wdmaud/entry.c:205:30: error: variable 'DeviceExtension' set but not used [-Werror=unused-but-set-variable]
```

**Comment by HBelusca** [comment]
> 
> 
> > What did you change in the ReactOS_LogOn.wav file ??
> > I changed the sample rate from 44.1khz to 48khz because VirtualBox AC97 driver wont play with 44.1khz

Huuh... Invalid answer. It plays occasionnaly (as mentioned by @BenNottelling), and, it plays OK on Windows 2003 installed on VBox with VBox AC97 driver.

**Comment by binarymaster** [comment]
> I changed the sample rate from 44.1khz to 48khz because VirtualBox AC97 driver wont play with 44.1khz

This could be true for VirtualBox 6.x branch. However 44.1 kHz sound should work without major problems on VirtualBox 5.x.

> and, it plays OK on Windows 2003 installed on VBox with VBox AC97 driver

WinXP / 2k3 have a mixer which converts any sample rate to appropriate one which is supported by the sound card driver.

**Comment by BenNottelling** [comment]
Hiding that GCC error can be done as seen here: 
https://github.com/reactos/reactos/commit/fa221ff1a455f097bfbbfd2ffc1e741c296fe2dd#diff-92D604F6ACB71A18F060F42F88B9FE83

**Comment by BenNottelling** [comment]
@Extravert-ir 
Fix for GCC error is adding this to https://github.com/reactos/reactos/blob/master/drivers/wdm/audio/legacy/wdmaud/wdmaud.h (which hides the error for an unused variable) 
```
#ifdef __GNUC__
#pragma GCC diagnostic ignored "-Wunused-variable"
#endif
```

File: drivers/wdm/audio/legacy/wdmaud/deviface.c [hunk 12]
```diff
@@ -49,10 +73,58 @@ DeviceInterfaceChangeCallback(
...
+    ULONG Length;
+    NTSTATUS Status;
 
     DPRINT1("DeviceInterfaceChangeCallback called %p\n", Event);
-    DbgBreakPoint();
+
+    if (IsEqualGUIDAligned(&Event->Event,
+                           &GUID_DEVICE_INTERFACE_ARRIVAL))
+    {
+        Length = (Event->SymbolicLinkName->Length);
```
[the remark this answers is not in the document]
Comment (spetreolle) [reply drivers/wdm/audio/legacy/wdmaud/deviface.c]: ... Nope.

**Comment by JoachimHenze** [comment]
I am so excited to see this PR janderwald.
Which of those defects will this PR have a chance to fix? CORE-9986? CORE-8726?
(CORE-8726? CORE-10118? CORE-9475?)

**Comment by Zero3K** [comment]
> 
> 
> I am so excited to see this PR janderwald.
> Which of those defects will this PR have a chance to fix? CORE-9986? CORE-8726?
> (CORE-8726? CORE-10118? CORE-9475?)

I think all except the first one will be fixed by this PR.

**Comment by janderwald** [comment]
> 9475

I assume it fixes https://jira.reactos.org/browse/CORE-9986

**Comment by Zero3K** [comment]
The audio now doesn't work after applying this PR (even after multiple reboots into Debug mode). A debug log is attached.

[rosdbg_10162209.log](https://github.com/reactos/reactos/files/3273664/rosdbg_10162209.log)

File: drivers/ksfilter/ks/swenum.c [hunk 13]
```diff
@@ -2029,6 +2038,7 @@ KsServiceBusEnumPnpRequest(
     LARGE_INTEGER Time;
     PDEVICE_RELATIONS DeviceRelation;
     PBUS_DEVICE_ENTRY DeviceEntry;
+	KIRQL OldLevel;
```
Comment (ThFabba) [inline drivers/ksfilter/ks/swenum.c]: Broken indent

File: drivers/ksfilter/ks/swenum.c [hunk 14]
```diff
@@ -2104,6 +2111,14 @@ KsServiceBusEnumPnpRequest(
             /* complete pending irps */
             KspCompletePendingIrps(DeviceEntry, STATUS_DEVICE_REMOVED);
 
+            /* free device extension */
+            FreeItem(ChildDeviceExtension);
+
+            /* FIXME this should only be done while installing the drivers */
```
Comment (ThFabba) [inline drivers/ksfilter/ks/swenum.c]: FYI, the right thing to do is to never delete PDOs in the first place, unless their parent FDO is removed.
See item 4 in https://docs.microsoft.com/en-us/windows-hardware/drivers/kernel/removing-a-device-in-a-bus-driver

File: media/inf/audio.inf [hunk 15]
```diff
@@ -1,4 +1,4 @@
-; AUDIO.INF
+﻿b; AUDIO.INF
```
Comment (ThFabba) [inline media/inf/audio.inf]: Unintentional `b` :p

File: drivers/wdm/audio/sysaudio/main.c [hunk 16]
```diff
@@ -203,13 +203,14 @@ SysAudio_AddDevice(
     }
 
     /* Load kmixer */
+#if 0
```
Comment (ThFabba) [inline drivers/wdm/audio/sysaudio/main.c]: A comment with any `#if 0` is usually nice

**Comment by Zero3K** [comment]
I am now getting a different compilation error in RosBE after applying it. It is as follows:

`[1/9325] Performing build step for 'host-tools'
ninja: no work to do.
[2/4280] Building C object drivers/ksfilter/ks/CMakeFiles/ks.dir/swenum.c.obj
FAILED: drivers/ksfilter/ks/CMakeFiles/ks.dir/swenum.c.obj 
C:\PROGRA~2\RosBE\i386\bin\gcc.exe  -DBUILDING_KS -DDBG=1 -DKDBG=1 -DREACTOS_BINARY_DIR=\"C:/reactos/output-MinGW-i386\" -DREACTOS_SOURCE_DIR=\"C:/reactos\" -DUSE_COMPILER_EXCEPTIONS -DWINVER=0x502 -D_COMDDK_ -D_M_IX86 -D_SEH_ENABLE_TRACE -D_SETUPAPI_VER=0x502 -D_USE_32BIT_TIME_T -D_USE_PSEH3=1 -D_WIN32_IE=0x600 -D_WIN32_WINDOWS=0x502 -D_WIN32_WINNT=0x502 -D_X86_ -D__REACTOS__ -D__i386__ -D_inline=__inline -Di386 -Wa,--compress-debug-sections -D__RELFILE__="&__FILE__[__FILE__[0] == '.' ? sizeof \"../\" - 1 : sizeof REACTOS_SOURCE_DIR]" -pipe -fms-extensions -fno-strict-aliasing -nostdinc -mstackrealign -Wold-style-declaration -Wdeclaration-after-statement -fdebug-prefix-map="C:\reactos"=ReactOS -gdwarf-2 -gstrict-dwarf -femit-struct-debug-detailed=none -feliminate-unused-debug-symbols -march=pentium -mtune=i686 -Werror -Wall -Wpointer-arith -Wno-char-subscripts -Wno-multichar -Wno-unused-value -Wno-unused-const-variable -Wno-maybe-uninitialized -O1 -fno-optimize-sibling-calls -fno-omit-frame-pointer -mpreferred-stack-boundary=3 -fno-set-stack-executable -Winvalid-pch -Werror=invalid-pch -Idrivers/ksfilter/ks -I../drivers/ksfilter/ks -I../sdk/include -I../sdk/include/psdk -I../sdk/include/dxsdk -Isdk/include -Isdk/include/psdk -Isdk/include/dxsdk -Isdk/include/ddk -Isdk/include/reactos -Isdk/include/reactos/mc -I../sdk/include/crt -I../sdk/include/ddk -I../sdk/include/ndk -I../sdk/include/reactos -I../sdk/include/reactos/libs   -include C:/reactos/output-MinGW-i386/drivers/ksfilter/ks/ks_pch.h -MMD -MT drivers/ksfilter/ks/CMakeFiles/ks.dir/swenum.c.obj -MF drivers/ksfilter/ks/CMakeFiles/ks.dir/swenum.c.obj.d -o drivers/ksfilter/ks/CMakeFiles/ks.dir/swenum.c.obj -c ../drivers/ksfilter/ks/swenum.c
../drivers/ksfilter/ks/swenum.c: In function 'KsServiceBusEnumPnpRequest':
../drivers/ksfilter/ks/swenum.c:2041:8: error: unused variable 'OldLevel' [-Werror=unused-variable]
../drivers/ksfilter/ks/swenum.c: At top level:
cc1.exe: error: unrecognized command line option "-Wno-unused-const-variable" [-Werror]
cc1.exe: all warnings being treated as errors
[3/4280] cmd.exe /C "cd /D C:\reactos\output-MinGW-i386\boot && C:\reactos\output-MinGW-i386\host-tools\fatten.exe C:/reactos/output-MinGW-i386/boot/efisys.bin -format 2880 EFIBOOT -boot C:/reactos/output-MinGW-i386/boot/freeldr/bootsect/fat.bin -mkdir EFI -mkdir EFI/BOOT -add C:/reactos/output-MinGW-i386/boot/environ/bootmgfw.efi EFI/BOOT/bootia32.efi"
[4/4280] Building C object drivers/network/afd/CMakeFiles/afd.dir/afd/listen.c.obj
[5/4280] Building C object drivers/network/afd/CMakeFiles/afd.dir/afd/read.c.obj
[6/4280] Building C object drivers/network/afd/CMakeFiles/afd.dir/afd/lock.c.obj
[7/4280] Building C object drivers/network/afd/CMakeFiles/afd.dir/afd/main.c.obj
ninja: build stopped: subcommand failed.`

**Comment by oleg-dubinskiy** [comment]
After removing `KIRQL OldLevel;` from `/drivers/ksfilter/ks/swenum.c:2041` and disabling usbaudio in `drivers/usb/CMakeLists.txt` and `boot/bootdata/txtsetup.sif`, it compiles successfully for me. And sound works correctly at least with AC97 VBox driver in VirtualBox 5.2.26, but still after two reboots, not after one. And it is necessary to replace portcls.sys on version from Win2k3 before AC97 installation, since with our portcls, after applying this PR, appears an exception and the system crashes to BSOD 0x1e during AC97 installation, unlike current master:
```
[7h
Entered debugger on last-chance exception (Exception Code: 0xc0000005) (Page Fault)
Memory at 0x018B0000 could not be written: Page not present.
kdb:> bt
Eip:
<ntoskrnl.exe:160456 (sdk/lib/rtl/unicode.c:644 (RtlInitUnicodeString))>
Frames:
<portcls.sys:1433 (drivers/wdm/audio/backpln/portcls/adapter.cpp:275 (PcRegisterSubdevice))>
<ALCXWDM.SYS:14973>
<ALCXWDM.SYS:14b1c>
<portcls.sys:5a4e (drivers/wdm/audio/backpln/portcls/irp.cpp:98 (PortClsPnp))>
<portcls.sys:5c90 (drivers/wdm/audio/backpln/portcls/irp.cpp:450 (PcDispatchIrp))>
<ntoskrnl.exe:725b0 (ntoskrnl/io/iomgr/irp.c:1286 (IofCallDriver))>
Couldn't access memory at 0xF70A1A50!
kdb:> cont

*** Fatal System Error: 0x0000001e
                       (0xC0000005,0x80960456,0xF70A1880,0x00000000)


Entered debugger on embedded INT3 at 0x0008:0x80955bb4.
kdb:> 
```
Nevertheless I hope that it may work on my real hardware too, at least with portcls.sys replaced. For now, sound doesn't work in ReactOS with my cards at all.

**Comment by Zero3K** [comment]
> 
> 
> After removing `KIRQL OldLevel;` from `/drivers/ksfilter/ks/swenum.c:2041` and disabling usbaudio in `drivers/usb/CMakeLists.txt` and `boot/bootdata/txtsetup.sif`, it compiles successfully for me. And sound works correctly at least with AC97 VBox driver in VirtualBox 5.2.26, but still after two reboots, not after one. And it is necessary to replace portcls.sys on version from Win2k3 before AC97 installation, since with our portcls, after applying this PR, appears an exception and the system crashes to BSOD 0x1e during AC97 installation, unlike current master:
> 
> ```
> �[7h�
> Entered debugger on last-chance exception (Exception Code: 0xc0000005) (Page Fault)
> Memory at 0x018B0000 could not be written: Page not present.
> kdb:> bt
> Eip:
> <ntoskrnl.exe:160456 (sdk/lib/rtl/unicode.c:644 (RtlInitUnicodeString))>
> Frames:
> <portcls.sys:1433 (drivers/wdm/audio/backpln/portcls/adapter.cpp:275 (PcRegisterSubdevice))>
> <ALCXWDM.SYS:14973>
> <ALCXWDM.SYS:14b1c>
> <portcls.sys:5a4e (drivers/wdm/audio/backpln/portcls/irp.cpp:98 (PortClsPnp))>
> <portcls.sys:5c90 (drivers/wdm/audio/backpln/portcls/irp.cpp:450 (PcDispatchIrp))>
> <ntoskrnl.exe:725b0 (ntoskrnl/io/iomgr/irp.c:1286 (IofCallDriver))>
> Couldn't access memory at 0xF70A1A50!
> kdb:> cont
> 
> *** Fatal System Error: 0x0000001e
>                        (0xC0000005,0x80960456,0xF70A1880,0x00000000)
> 
> 
> Entered debugger on embedded INT3 at 0x0008:0x80955bb4.
> kdb:> 
> ```
> 
> Nevertheless I hope that it may work on my real hardware too, at least with portcls.sys replaced. For now, sound doesn't work in ReactOS with my cards at all.

1. Thanks for letting me know.
2. Hopefully someone can fix that crash soon.

**Comment by JoachimHenze** [comment]
Please do not merge to master until unrelated regression https://jira.reactos.org/browse/CORE-16157 got a fix.

**Comment by BenNottelling** [comment]
Regression of [CORE-16157](https://jira.reactos.org/browse/CORE-16157) has been resolved in https://github.com/reactos/reactos/commit/76e00fab7e959d4b4acd17dababc02f26c2208ef

**Comment by Zero3K** [comment]
I wonder if https://jira.reactos.org/browse/CORE-16190 could also be fixed by this modifying this patch.

**Comment by Zero3K** [comment]
What do I change to fix the second one?

**Comment by Zero3K** [comment]
I tried applying the newest changes and now the audio doesn't work at all.

**Comment by Zero3K** [comment]
When running OrgView.exe, I am getting this in the debug log:
(drivers/ksfilter/swenum/swenum.c:314) SwDispatchCreate 0
(drivers/ksfilter/ks/swenum.c:1927) KsServiceBusEnumCreateRequest IRP B443EBC8 Name \{cd171de3-69e5-11d2-b56d-0000f8754380}&{9b365890-165f-11d0-a195-0020afd156e4}
(drivers/ksfilter/swenum/swenum.c:328) SwDispatchCreate 104
(drivers/wdm/audio/backpln/portcls/pin_wavecyclic.cpp:1241) CPortPinWaveCyclic::Init Status c0000272 PinId 0 Capture 0
(drivers/wdm/audio/backpln/portcls/pin_wavecyclic.cpp:1241) CPortPinWaveCyclic::Init Status c0000272 PinId 0 Capture 0

When I run ReactOS Media Player and play a .wav, the CPortPinWave::Init Status part in the debug log changes to the following:

CPortPinWaveCyclic::Init Status 0 PinId 0 Capture 0

**Comment by BenNottelling** [comment]
(From Audacity) I don't think we're detecting the supported sample rates correctly? 
image
From [Wikipedia](https://en.wikipedia.org/wiki/AC%2797):
`AC'97 supports a 96 kHz sampling rate at 20-bit stereo resolution and a 48 kHz sampling rate at 20-bit stereo`

I'm not sure which revision AC'97 that is though. 
`Revision 1.x` Only supports a fixed 48 kHz
`Revision 2.1` Brought in support for multiple channels & and a sample rate range 

None of the revisions mention supporting anything more than 48 kHz.
I've found a AC'97 chip on Texas Instrument's website, (rev 2.1) and it specifically says
```
High-Quality Sample Rate Conversion from 4 kHz 
to 48 kHz in 1-Hz Increments
```

I know for a fact that VLC is supposed to resample audio to whatever the device supports, which would explain why it doesn't play audio if it's detecting the wrong sample rate.

**Comment by BenNottelling** [comment]
`araw decoder debug: samplerate:48000Hz channels:2 bits/sample:16`
Nope! VLC is resampling 48 kHz so that's not the issue

**Comment by Zero3K** [comment]
Can someone take a look at the DriverProc and WodMessage functions of wdmaud.drv? I think one or both of them is causing portcls.sys to lag the audio/freeze programs/have no audio playing.

**Comment by jeditobe** [comment]
Is this PR a bit rotten now?

**Comment by Zero3K** [comment]
The audio playback issue in VirtualBox was fixed in v6.0.14 of it. Now only the issues with portcls.sys remain. So, I think that this PR should be fixed and merged soon since it helps fix issues on real hardware and VMs.

**Comment by Zero3K** [comment]
> 
> 
> Is this PR a bit rotten now?

Seems to be.

**Comment by blackcrack** [comment]
@ThFabba 
*looking on this code and want Remind you* :)

**Comment by Zero3K** [comment]
I wonder what happened to @janderwald.

**Comment by Zero3K** [comment]
Here's the fixed patches provided by Illen (on Discord):

[audio_pr_fixed.zip](https://github.com/reactos/reactos/files/4013607/audio_pr_fixed.zip)

**Comment by Zero3K** [comment]
Welcome back @janderwald :smiley:

**Comment by Saibamen** [comment]
+1 to wanted feature :)
Could you please update PR title?

**Comment by Zero3K** [comment]
> 
> 
> +1 to wanted feature :)
> Could you please update PR title?

Why? This is a PR that fixes lots of audio related issues.

**Comment by Saibamen** [comment]
It could be just `[AUDIO] Fix sound after first reboot`

**Comment by Zero3K** [comment]
Here's a program to help test the audio code;

https://jira.reactos.org/secure/attachment/53685/OrgView.exe

**Comment by Zero3K** [comment]
Please fix it so it can be compiled by MinGW and MSVC.

**Comment by Extravert-ir** [comment]
> Could you please update PR title?

Could you please stop writing useless comments? You're only generating spam here

**Comment by Zero3K** [comment]
> 
> 
> > Could you please update PR title?
> 
> Could you please stop writing useless comments? You're only generating spam here

Okay. I hope that it gets fixed eventually.

File: dll/win32/streamci/streamci.spec [hunk 17]
```diff
@@ -1 +1,2 @@
 7 stdcall StreamingDeviceSetupW(ptr ptr ptr long)
+@ stdcall SwEnumCoInstaller(long ptr long long)
```
Comment (binarymaster) [inline dll/win32/streamci/streamci.spec:2]: image

File: drivers/usb/usbccgp/fdo.c [hunk 18]
```diff
@@ -190,6 +190,10 @@ FDO_CreateChildPdo(
     /* Create pdo for each function */
     for(Index = 0; Index < FDODeviceExtension->FunctionDescriptorCount; Index++)
     {
+		if (FDODeviceExtension->FunctionDescriptor[Index].NumberOfInterfaces == 0)
+		{
+			continue;
+		}
```
Comment (binarymaster) [inline drivers/usb/usbccgp/fdo.c:196]: Indentation problems. Also I think `usbccgp` fixes deserve a separate PR :)

**Comment by Zero3K** [comment]
This PR no longer applies. I hope that it can be fixed soon.

**Comment by BenNottelling** [comment]
> This PR no longer applies. I hope that it can be fixed soon.

6a6d1124cf69e5588588bc7e397598f6
